File: src/main/java/com/github/takemikami/intellij/plugin/externallinters/CommandUtil.java
package com.github.takemikami.intellij.plugin.externallinters;

import com.intellij.openapi.diagnostic.Logger;
import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Map;

/**
 * Command Utility for Inspection.
 */
public class CommandUtil {

  private static final Logger LOG = Logger.getInstance(CommandUtil.class);

  public static String runCommand(String[] cmd, String workingdir, Map<String, String> env)
      throws IOException {
    return CommandUtil.runCommand(cmd, workingdir, env, null);
  }

  /**
   * Execute process.
   *
   * @param cmd        command and parameters
   * @param workingdir working directory path
   * @param env        environment variables
   * @param stdin      standard in text
   * @return standard output string
   * @throws IOException execution failed
   */
  public static String runCommand(String[] cmd, String workingdir, Map<String, String> env,
      String stdin) throws IOException {
    ProcessBuilder pb = new ProcessBuilder(cmd);
    pb.directory(new File(workingdir));
    if (env != null) {
      pb.environment().putAll(env);
    }
    Process p = pb.start();
    if (stdin != null) {
      OutputStream out = p.getOutputStream();
      out.write(stdin.getBytes(StandardCharsets.UTF_8));
      out.flush();
      out.close();
    }
    try {
      p.waitFor();
    } catch (InterruptedException ex) {
      LOG.error("command " + cmd[0] + " error: ", ex);
      return "";
    }
    String stderr = new String(p.getErrorStream().readAllBytes(), StandardCharsets.UTF_8);
    LOG.debug("command " + cmd[0] + " stderr: ", stderr);
    return new String(p.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
  }

  /**
   * Return command path.
   *
   * @param cmd command name
   * @return command path
   */
  public static String findCommandPath(String cmd) {
    String out;
    try {
      out = CommandUtil.runCommand(new String[]{"which", cmd}, "/", null).strip();
      if (out.strip().length() == 0) {
        return null;
      }
    } catch (IOException ex) {
      LOG.error("command which " + cmd + " error: ", ex);
      return null;
    }
    if (!Files.exists(Paths.get(out))) {
      return null;
    }
    return out;
  }
}

File: src/main/kotlin/com/github/takemikami/intellij/plugin/pyprojectlinters/CommandUtil.kt
package com.github.takemikami.intellij.plugin.pyprojectlinters

import java.io.File
import java.io.IOException
import java.io.OutputStream
import java.nio.charset.StandardCharsets

/**
 * Command Utility for Inspection.
 */
class CommandUtil {
    companion object {
        /**
         * Execute process.
         *
         * @param cmd        command and parameters
         * @param workingdir working directory path
         * @param env        environment variables
         * @param stdin      standard in text
         * @return standard output string
         * @throws IOException execution failed
         */
        @Throws(IOException::class)
        fun runCommand(
            cmd: Array<String?>,
            workingdir: String?,
            env: MutableMap<String?, String?>?,
            stdin: String?,
        ): String? {
            val pb = ProcessBuilder(*cmd)
            pb.directory(File(workingdir))
            if (env != null) {
                pb.environment().putAll(env)
            }
            val p = pb.start()
            if (stdin != null) {
                val out: OutputStream = p.getOutputStream()
                out.write(stdin.toByteArray(StandardCharsets.UTF_8))
                out.flush()
                out.close()
            }
            try {
                p.waitFor()
            } catch (ex: InterruptedException) {
                return null
            }
            val stderr = String(p.getErrorStream().readAllBytes(), StandardCharsets.UTF_8)
            return String(p.getInputStream().readAllBytes(), StandardCharsets.UTF_8)
        }
    }
}
